### PR TITLE
reconcile service resource

### PR DESCRIPTION
**Pull request by ashleyschuett**
Signed-off-by: Ashley Schuett <aschuett@redhat.com>

**What this PR does / why we need it**:
Changed reconciling of labels and annotations to be more general like the other resources. Mainly refactored the function to be more readable and also removed duplicate unit tests since they where in two separate files. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

```release-note
Reconcile service resource
```

### CODE REVIEW

**Review by acardace — COMMENTED** [review]
Why did you decide to look for resource changes in this way? you deleted the snippet
```
// First check if there's anything to do.
	if objectMatchesVersion(&cachedService.ObjectMeta, version, imageRegistry, id, r.kv.GetGeneration()) {
		// spec and annotations are already up to date. Nothing to do
		return patchOps, false, nil
	}
```

which would first check the generation before doing any processing, by getting rid of that you have the reconciling loop do a bit more than what it would do before in the case no reconciliation is needed.

I'm not saying you should have used the expected generations approach we use for other resources as clearly both approaches work and it's a tradeoff between CPU and memory (generations store more info), but maybe we should keep the snippet above to skip processing this function if the version has not changed?

**Comment by ashleyschuett** [comment]
> Why did you decide to look for resource changes in this way? you deleted the snippet

> // First check if there's anything to do.
	if objectMatchesVersion(&cachedService.ObjectMeta, version, imageRegistry, id, r.kv.GetGeneration()) {
		// spec and annotations are already up to date. Nothing to do
		return patchOps, false, nil
	}

> which would first check the generation before doing any processing, by getting rid of that you have the reconciling loop do a bit more than what it would do before in the case no reconciliation is needed.

> I'm not saying you should have used the expected generations approach we use for other resources as clearly both approaches work and it's a tradeoff between CPU and memory (generations store more info), but maybe we should keep the snippet above to skip processing this function if the version has not changed?

@acardace APIService does not have a generation field which is why I am doing a compare of the fields we expect to not change. The deleted snipit is actually just checking that labels/annotations are as we would expect which is now done in `resourcemerge.EnsureObjectMeta(modified, existingCopy, desired)`

**Comment by acardace** [comment]
> > Why did you decide to look for resource changes in this way? you deleted the snippet
> 
> > // First check if there's anything to do.
> > if objectMatchesVersion(&cachedService.ObjectMeta, version, imageRegistry, id, r.kv.GetGeneration()) {
> > // spec and annotations are already up to date. Nothing to do
> > return patchOps, false, nil
> > }
> 
> > which would first check the generation before doing any processing, by getting rid of that you have the reconciling loop do a bit more than what it would do before in the case no reconciliation is needed.
> 
> > I'm not saying you should have used the expected generations approach we use for other resources as clearly both approaches work and it's a tradeoff between CPU and memory (generations store more info), but maybe we should keep the snippet above to skip processing this function if the version has not changed?
> 
> @acardace APIService does not have a generation field which is why I am doing a compare of the fields we expect to not change. The deleted snipit is actually just checking that labels/annotations are as we would expect which is now done in `resourcemerge.EnsureObjectMeta(modified, existingCopy, desired)`

Gotcha, now I see :+1:

**Comment by acardace** [comment]
/lgtm

But maybe https://github.com/kubevirt/kubevirt/pull/5459/commits/d02040e58913a530e03de8229fc598257a435ba6 and https://github.com/kubevirt/kubevirt/pull/5459/commits/a4071b5174e4ad0e4efb804943b70bc5c806ff5d belong to a separate PR? Why have they been introduced in this one?

File: pkg/virt-operator/resource/apply/core.go [hunk 1]
```diff
@@ -300,81 +329,75 @@ func (r *Reconciler) createOrUpdateComponentsWithCertificates(queue workqueue.Ra
...
+func hasImmutableFieldChanged(service, cachedService *corev1.Service) bool {
+	typeSame := false
+	deleteAndReplace := false
 
-			// If both the cached service and the desired service have ClusterIPs set
-			// that are not equal, our only option is to delete/replace because ClusterIPs
-			// are not mutable
-			deleteAndReplace = true
-		}
+	requiredIsEmpty := len(service.Spec.Type) == 0
```
[the remark this answers is not in the document]
Comment (ashleyschuett) [reply pkg/virt-operator/resource/apply/core.go]: you're right. The reason I left it was from fiddling with another idea. which I forgot to point out / comment on. I'm interested in peoples thoughts about this https://github.com/kubevirt/kubevirt/pull/5459/files#diff-971364d0a5491acdd4a676a6d6e98c316e574b9479eebcc4773a282ac1c785e9R67-R68

**Comment by ashleyschuett** [comment]
/hold 

https://github.com/kubevirt/kubevirt/pull/5459/files#diff-971364d0a5491acdd4a676a6d6e98c316e574b9479eebcc4773a282ac1c785e9R67-R68

should this trigger a delete and replace? If the "desired" service type and current service type are the same even if not of type ClusterIP that should be okay. I think this test should be removed

File: pkg/virt-operator/resource/apply/core.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (acardace) [reply pkg/virt-operator/resource/apply/core.go]: When I first looked at it I thought 'it must have been there for a reason' but honestly I do not understand it, one thing you can look at is the git-blame, maybe the commit msg has some insights.

**Comment by fgimenez** [comment]
@ashleyschuett: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by ashleyschuett** [comment]
> https://github.com/kubevirt/kubevirt/pull/5459/files#diff-971364d0a5491acdd4a676a6d6e98c316e574b9479eebcc4773a282ac1c785e9R67-R68

> should this trigger a delete and replace? If the "desired" service type and current service type are the same even if not of type ClusterIP that should be okay. I think this test should be removed

@davidvossel I see you wrote the original test for this https://github.com/kubevirt/kubevirt/blame/d575b867afd043173ba385001dc85c846623c60d/pkg/virt-operator/install-strategy/create_test.go#L361-L383

 I think it makes sense that we only reconcile ClusterIP service. however by patching if the types are equal but the desired type if not of type ClusterIP we end up in an infinite reconciliation loop. Is it okay to change this behavior and change this test so that `deleteAndReplace` should be `false`?

**Comment by davidvossel** [comment]
> should this trigger a delete and replace? If the "desired" service type and current service type are the same even if not of type ClusterIP that should be okay. I think this test should be removed

I only tested that applying updates works with ClusterIP type, and that required me to understand the relationship between the dynamically set clusterIP and our cached resource. 

I don't know how the other service types work. If they have defaults being dynamically set then we run the risk that our cached service won't match the one in the install strategy due to the defaulting values. That's why i took the conservative approach of delete/replace

**Comment by ashleyschuett** [comment]
@davidvossel thanks for the explanation. sounds good, I have removed my comment. 

/unhold

**Review by stu-gott — COMMENTED** [review]
/approve

**Pull request merged**